### ISSUE TITLE
Form errors with JSON API

### ISSUE THREAD

**Issue by ken92** [erikras/redux-form#2569]
I've been looking around for a solution to this but I haven't found one.  If one exists already, sorry about the double-posting.

I'm working with an API that returns JSON in this kind of format:
`{'is_successful' => 1, 'data' => {some data here}}`
or
`{'is_successful' => 0, 'error' => "Error string"}`

I want to display the errors I'm getting back from the database when I get them, but whenever I think I might have something, I just get "uncaught in promise" errors.

This is what I have right now, with a lot of placeholder data filling in for useful things until I figure out how everything works:
The element that contains the redux-form is put on the page with this:
`<LoginForm submitFunction={this.validateAndDoAThing.bind(this)} />`

The function being passed in:
```
validateAndDoAThing = (values, dispatch) => {
		return new Promise((resolve, reject) => {
			console.log("values: ", values);
			dispatch(loginUser(values.email, values.password)).then((response) => {
				let data = response.payload.data;

				console.log("data: ", data);

				if(response.payload.status != 200 || data.is_successful == 0) {
					dispatch(testLoginError(response.payload));
					reject({password: 'Wrong password', _error: 'Login failed!'});
				} else {
					dispatch(testLoginSuccess(response.payload)); 
					resolve();
				}
			});
		});
	};
```

The login form:
```
import React from 'react';
import {Field, reduxForm} from 'redux-form';

const validate = values => {
	const errors = {};
	if (!values.email)
		errors.email = 'Please enter your email';
	else if (!isValidEmail(values.email))
		errors.email = 'Invalid email address!';
	
	if (!values.password)
		errors.password = 'Please enter your password';
	
	return errors;
}

const renderField = ({ input, label, type, meta: { touched, error, warning } }) => (
	<div>
		<input {...input} placeholder={label} type={type}/>
		{touched && ((error && <span>{error}</span>) || (warning && <span>{warning}</span>))}
	</div>
)

const LoginForm = (props) => {
	const {handleSubmit, error} = props;
	return (
		<form onSubmit={handleSubmit(props.submitFunction.bind(this))}>
			<Field name="email" type="email" label="email" component={renderField} />
			<Field name="password" type="password" label="password" component={renderField} />

			{error && <div>{error}</div>}

			<p className="add-margin-bottom-30"><a href="#">Forgot Password</a></p>
			<button type="submit">Sign In</button>
			<button className="fb-button"><img src="images/fb-small-white.png" />Sign In with Facebook</button>
			<p className="text-center"><a href="javascript:void(0);" onclick="showCreateAccountModal();">Create Account</a></p>
		</form>
	)
}

export default reduxForm({
	form: 'loginForm',
	validate
})(LoginForm)
```

Right now, the test success and error functions are being called appropriately, but no errors are showing up for the user on the form.  Instead, I just get this in the console:
`Uncaught (in promise):
_error: "Login failed!"
password: "Wrong password"`

I've been trying to figure out how to get this to work for two days, now, so any help would be appreciated!  Thanks.

**Comment by pcgilday** [comment]
Have you tried throwing a [SubmissionError](http://redux-form.com/6.5.0/docs/api/SubmissionError.md/)?

**Comment by ken92** [comment]
Yeah, and I just get a console error with nothing displaying

**Comment by pcgilday** [comment]
Would you mind providing the code snippet using `SubmissionError`, and the error you get in console (unless it is the same uncaught in promise error)?

I don't believe you should get the uncaught promise error if `SubmissionError` is used; however, if it isn't used, that error makes sense. See [these lines here](https://github.com/erikras/redux-form/blob/master/src/handleSubmit.js#L42), `handleSubmit` re-throws the error if it is not an instance of `SubmissionError` and `onSubmitFailed` was not set, as it expects you to handle the thrown error in that scenario.

**Comment by ken92** [comment]
Sure thing.

```
validateAndDoAThing(values, dispatch) {
		return new Promise((resolve, reject) => {
			dispatch(loginUser(values.email, values.password))
				.then((response) => {
					let data = response.payload.data;

					if(response.payload.status != 200 || data.is_successful == 0) {
						throw new SubmissionError({email: data.error, _error: data.error});
					} else {
						dispatch(testLoginSuccess(response.payload)); 
						resolve();
					}
			});
		});
	}
```

The error:
```
Uncaught (in promise):
errors: [Object]
message: "Submit Validation Failed"
name: "SubmissionError"
stack: "SubmissionError: Submit Validation Failed↵ 
```

**Comment by pcgilday** [comment]
I am guessing you are using middleware that makes `dispatch` return a promise? If so, you don't need or want to wrap that in a new promise, something like this would be clearer, and should work. 

```js
validateAndDoAThing(values, dispatch) {
    return dispatch(loginUser(values.email, values.password))
        .then((response) => {
	    let data = response.payload.data;

	    if(response.payload.status != 200 || data.is_successful == 0) {
	        throw new SubmissionError({email: data.error, _error: data.error});
	    } else {
	        dispatch(testLoginSuccess(response.payload)); 
	        resolve();
            }
        });
}
```

If `dispatch` is not returning a promise, you would do something like this. 

```js
validateAndDoAThing(values, dispatch) {
    return Promise.resolve(dispatch(loginUser(values.email, values.password)))
        .then((response) => {
	    let data = response.payload.data;

	    if(response.payload.status != 200 || data.is_successful == 0) {
	        throw new SubmissionError({email: data.error, _error: data.error});
	    } else {
	        dispatch(testLoginSuccess(response.payload)); 
	        resolve();
            }
        });
}
```

TLDR: you need to handle the response and throw the the error within the `.then` function attached to the returned promise.

**Comment by ken92** [comment]
That worked, thanks so much!

Maybe this middleware use case could be included on the examples page?  It hadn't even occurred to me that the middleware would be causing my issues.

**Comment by pcgilday** [comment]
No problem, happy to help! 

I'm not sure it belongs in the examples, as this was an issue with how promise chains work. Promises aren't meant to be nested like that, and therefore don't "bubble up" errors to the containing promise chain. Since ReduxForm chains a `.then` to the promise you returned, that is the promise chain that needs to return needed values or throw any errors.
